umbraco/Umbraco-CMS
# New Backoffice: Health check controller

## Details
- New endpoints for the Health Check dashboard;
  - <code>/health-check-<b>group</b></code> - gets all health checks in a paginated grouped list;
    - The idea here is to get all the groups so that we can display them in a grid and for every group call the other endpoint ⬇️ to load and show the individual health check result(s) (_as seen on the image_);
    - By having this logic separately, we will avoid blocking the endpoint result in case a single health check is taking more time than the others;
    - That way each group can have a loading wheel or general loading bar for all the health checks so that the user is aware that something is happening and they don't resort to reloading the website and making the same request for all health checks once again.
  - <code>/health-check-<b>group</b>/{name}</code> -  gets a health check group found with the same name and contains each health check result(s) or returns a not found result;
    - Heps to get the overall result from running all health checks in a group;
    - ❗ The case is ignored when trying to find a group with the same name;
    - ❗ We don't have ids/keys for the groups, so the only way to request a health check group at this time is by groupName.
  - `health-check/execute-action` - executes a given action from a health check or retuns a _Bad Request_;
    - If a health check provides actions that can be executed to fix a health check when it fails, then we can use the endpoint to rectify the issue;
    - The request body is in the form of a HealthCheckAction:
```json
{
  "healthCheckKey": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "alias": "string",
  "name": "string",
  "description": "string",
  "valueRequired": true,
  "providedValue": "string",
  "providedValueValidation": "string",
  "providedValueValidationRegex": "string"
}
```
</br>

![image](https://user-images.githubusercontent.com/21998037/206442052-bbbb5283-c96b-4e34-8a81-e0232ce98404.png)

---
_Note ❗ : the `resultType` of a health check can be :_
- "Success" ✔️ 
- "Error" ❌ 
- "Warning" ⚠️ 
- "Info" ℹ️ 
---
A quick change of <em><code>LanguageViewModelsMapDefinition.cs</code></em> is included in this PR since I noticed it while implementing a mapper for the health checks:
- It is related to removing a few lines for null and type checks which are unnecessary since we instantiate the target before calling the `Map()` where the checks were. Meaning that we already make sure that those will never be true. 😉 

</br>

## Test
- The easiest to test is using Swagger - visit `localhost/umbraco/swagger` as the majority of endpoints are Get requests;
- `/health-check-group`
  - Execute with the default skip and take values or modify them as you like;
  - Check the response with the health check groups that we get in the "old" backoffice - we should have the same amount of groups and the health checks should be grouped as before;
  - Try adding a custom health check, as shown in the code example of _[general checks section](https://docs.umbraco.com/umbraco-cms/extending/health-check#general-checks)_ and make sure it is included in the response;
  - Try to disable a health check and verify that it is NOT included in the response;
    - You can get their ids/keys from _[here](https://docs.umbraco.com/umbraco-cms/extending/health-check#built-in-checks)_;
    - _Note:_ You have to disable all checks from a group not to include it in the response.

_ex:_
```json
"Umbraco": {
  "CMS": {
    "HealthChecks": {
      "DisabledChecks": [
        {
          "Id": "D0F7599E-9B2A-4D9E-9883-81C7EDC5616F"
        }
      ]
    }
  }
}
```  

- `/health-check-group/{name}`
  - Edit the `name` - ex: `Permissions`;  
  - Make sure the details of the group are correct (you can compare them to the response from the precious endpoint ⬆️ );
  - Make sure that the information in the result(s) (`"results": []`) are identical to the ones from the "old" backoffice.

- `health-check/execute-action` - _POST_ request
  - You can add the custom health check shown in the code example (the whole `RobotsTxt` class) in _[general checks section](https://docs.umbraco.com/umbraco-cms/extending/health-check#general-checks)_;
    - It adds an action that you can execute, which will add a robots.txt file and therefore satisfy the health check;
  - Check the returned response from clicking the action button of the SEO > Robots.txt healthcheck in the "old" backoffice (it will make a call to the old ExecuteAction endpoint);
  - Delete the created robots.txt file;
  - Go to `localhost/umbraco/swagger` and try the `health-check/execute-action` endpoint with the following body:
```json
{
  "healthCheckKey": "3a482719-3d90-4bc1-b9f8-910cd9cf5b32",
  "alias": "addDefaultRobotsTxtFile",
  "name": "Do the thing!",
  "description": "Clicking button will do the thing and fix your site",
  "valueRequired": false,
  "providedValue": null,
  "providedValueValidation": null,
  "providedValueValidationRegex": null
}
```
-
  - Compare the response body with the one from the "old" backoffice;
  - Verify that the response from `/health-check-group/SEO` no longer contains any actions and that the `resultType` changed from `"Error"` to `"Success"`.

</br>

## Appendix
### `/health-check-group` endpoint response - _Paged HealthCheckGroup-s_
```json
{
  "total": 0,
  "items": [
    {
      "name": "string",
      "checks": [
        {
          "key": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
          "name": "string",
          "description": "string"
        }
      ]
    }
  ]
}
```


### `/health-check-group/{name}` endpoint response - _HealthCheckGroupWithResult_
```json
{
  "name": "string",
  "checks": [
    {
      "key": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "name": "string",
      "description": "string",
      "results": [
        {
          "message": "string",
          "resultType": "Success",
          "actions": [
            {
              "healthCheckKey": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
              "alias": "string",
              "name": "string",
              "description": "string",
              "valueRequired": true,
              "providedValue": "string",
              "providedValueValidation": "string",
              "providedValueValidationRegex": "string"
            }
          ],
          "readMoreLink": "string"
        }
      ]
    }
  ]
}
```
OR (_when not found_)
```json
{
  "statusCode": 0
}
```

### `health-check/execute-action` endpoint response - _HealthCheckResult_
```json
{
  "message": "string",
  "resultType": "Success",
  "actions": [
    {
      "healthCheckKey": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "alias": "string",
      "name": "string",
      "description": "string",
      "valueRequired": true,
      "providedValue": "string",
      "providedValueValidation": "string",
      "providedValueValidationRegex": "string"
    }
  ],
  "readMoreLink": "string"
}
```
OR (_ProblemDetails as Bad Request_)
```json
{
  "type": "string",
  "title": "string",
  "status": 0,
  "detail": "string"
}
```
